### PR TITLE
task288 - add dyndns tasks,test,handler

### PR DESCRIPTION
**Pull request by fnateghi**
Pull Request Description:

This pull request introduces a new task for managing Dynamic DNS (DynDNS) settings in OPNsense, along with associated handlers and tests. The changes include the ability to remove and update general DynDNS settings, apply account-specific configurations, and restart the DynDNS service automatically. Additionally, comprehensive tests have been added to ensure proper configuration handling.

Key updates:

DynDNS task for applying general and account-specific settings.
New handler for restarting the DynDNS service.
Tests to validate XML configurations for DynDNS.

### CODE REVIEW

**Comment by zerwes** [comment]
closes #95

File: tasks/dyndns.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
+---
+
+- name: DynDNS - remove DynDNS
+  delegate_to: localhost
+  community.general.xml:
+    path: "{{ local_config_path }}"
+    xpath: /opnsense/OPNsense/DynDNS{{ item }}
```
Comment (zerwes) [inline tasks/dyndns.yml:7]: I would leave this a task for `opn_unset`
But if you prefer to do it in the dyndns task, you should just expect a list of UUIDs in the VAR and then use
```yaml
xpath: "/opnsense/OPNsense/DynDNS/accounts/account[@uuid='{{ item }}']
```
to remove accounts

File: test/test-dyndn.yml
Comment (zerwes) [inline test/test-dyndn.yml:1]: Hello @fnateghi ... testing is great :+1: 
But ... Ough ... I think you have misunderstood the testing completely 
You should test your task(s) and not write a complete playbook
That is why your wrong condition in line 40 in the task file was not breaking the test.
The test should not test himself, but should test your task and finde errors like the wrong condition above.

you must register your new test tag
```diff
diff --git a/test/test.yml b/test/test.yml
index 2e8a39b..d071f55 100644
--- a/test/test.yml
+++ b/test/test.yml
@@ -44,6 +44,7 @@
         - ipsec
         - dnsserver
         - openvpn
+        - dyndns
       when:
         - test | default(_testtask) == _testtask
     - ansible.builtin.meta: flush_handlers
```
and maybe add the missing handlere as a fake one here too
and then  you define your VAR files for the tests as `test/dyndns-test*.yml`, optional some starting xml as `test/dyndns-test*.xml` and the expected result as `test/dyndns-test*-expect.xml`
and then you can run your test in the test directory using
`ansible-playbook -e test=dyndns test.yml`

File: tasks/dyndns.yml [hunk 2]
```diff
@@ -0,0 +1,41 @@
+---
+
```
Comment (zerwes) [inline tasks/dyndns.yml:2]: adding a comment with a example of the VARs here would be great

**Review by zerwes — CHANGES_REQUESTED** [review]

File: tasks/dyndns.yml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+- name: DynDNS -  Update general DynDNS settings
+  delegate_to: localhost
+  community.general.xml:
+    path: "{{ local_config_path }}"
+    xpath: "/opnsense/OPNsense/DynDNS/general/{{ item.key }}"
+    value: "{{ item.value }}"
+    state: present
+    pretty_print: true
+  notify: register dyndns
+  loop: "{{ opn_dyndns_general }}"
```
Comment (zerwes) [inline tasks/dyndns.yml:23]: and I would prefer to use `with_items' here
as in all other tasks

File: tasks/dyndns.yml [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    xpath: "/opnsense/OPNsense/DynDNS/accounts/account[@uuid='{{ item.0.uuid }}']/{{ item.1.key }}"
+    value: "{{ item.1.value }}"
+    state: present
+    pretty_print: true
+  notify: register dyndns
+  with_subelements:
+    - "{{ opn_dyndns_accounts }}"
+    - settings
+  when:
+    - opn_dyndns_accounts_ng is defined
```
Comment (zerwes) [inline tasks/dyndns.yml:40]:
```diff
-    - opn_dyndns_accounts_ng is defined
+    - opn_dyndns_accounts is defined
```

File: tasks/dyndns.yml [hunk 5]
```diff
@@ -0,0 +1,77 @@
+---
+#
+#opn_dyndns_general:
```
Comment (zerwes) [inline tasks/dyndns.yml:3]: pls fix the lint errors first ...

**Comment by fnateghi** [comment]
Dear @zerwes

Thank you for your prompt response and feedback. I encountered a small issue with my Git settings, which prevented me from committing the requested configuration initially. However, I have now modified the task and added the handler and the config example in the task as requested.

That said, I believe there might still be a minor issue in my code, though I'm not entirely sure what is wrong or where the problem lies. When I run the test playbook, I receive the following error message:


`TASK [fail due to diff between result and expected state for dyndns-test] ******************************************************************************************************************
task path: /.../opnsense/test/testsimpletaskgeneric.yml:46
fatal: [localhost]: FAILED! => {
    "changed": false,
    "msg": "fail due to diff between result and expected state for dyndns-test"
}`

the result should be the enrire settings, but in the *expect.xml file i can only see the general part of the DynDNS settings. The account part is missing somehow. 
Could you please assist me in identifying where I might have gone wrong?
thanks

**Comment by zerwes** [comment]
> Dear @zerwes
> 
> Thank you for your prompt response and feedback. I encountered a small issue with my Git settings, which prevented me from committing the requested configuration initially. However, I have now modified the task and added the handler and the config example in the task as requested.
> 
> That said, I believe there might still be a minor issue in my code, though I'm not entirely sure what is wrong or where the problem lies. When I run the test playbook, I receive the following error message:
> 
> `TASK [fail due to diff between result and expected state for dyndns-test] ****************************************************************************************************************** task path: /.../opnsense/test/testsimpletaskgeneric.yml:46 fatal: [localhost]: FAILED! => { "changed": false, "msg": "fail due to diff between result and expected state for dyndns-test" }`
> 
> the result should be the enrire settings, but in the *expect.xml file i can only see the general part of the DynDNS settings. The account part is missing somehow. Could you please assist me in identifying where I might have gone wrong? thanks

Hello @fnateghi 
I have already pointed you th the problem in the first review
https://github.com/Rosa-Luxemburgstiftung-Berlin/ansible-opnsense/pull/97#discussion_r1756788504

**Comment by zerwes** [comment]
so it might be desirable to iterate all my comments in the code review one by one

**Comment by fnateghi** [comment]
Sorry @zerwes. I will fix this and commit again.

**Comment by fnateghi** [comment]
I hope this time is everything correct :)

File: tasks/dyndns.yml [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    value: "{{ item.1.value }}"
+    state: present
+    pretty_print: true
+  notify: restart dyndns
+  with_subelements:
+    - "{{ opn_dyndns_accounts }}"
+    - settings
+  when:
+    - opn_dyndns_accounts is defined
+    - item.1.key != 'enable'
```
Comment (zerwes) [inline tasks/dyndns.yml:78]: :question: 
why do you exclude the enable tag here? how do you handle it?

File: test/test.yml [hunk 7]
```diff
@@ -26,6 +26,10 @@
     - name: stop ipsec  # TODO: test this action; use community.general.xml and add a tag to the resulting xml
       debug:
         msg: fake handler - stop ipsec
+    - name: restart dyndns
+      debug:
+        msg: fake handler - restart dyndns
+
```
Comment (zerwes) [inline test/test.yml:32]: pls, do not ad empty lines where there was none was before ...

File: tasks/dyndns.yml (hunk 6 above)
Comment (zerwes) [reply tasks/dyndns.yml:78]:
```diff
-    - item.1.key != 'enable'
```

**Review by zerwes — CHANGES_REQUESTED** [review]
> I hope this time is everything correct :)

nope, sadly not

File: test/dyndns-test.xml [hunk 8]
```diff
@@ -0,0 +1,36 @@
+<?xml version='1.0' encoding='UTF-8'?>
+<opnsense>
+  <system>
+  </system>
+  <OPNsense>
```
Comment (zerwes) [inline test/dyndns-test.xml:5]: again: it seems you did not understand thze scope of a test
your starting and expected xml are identic :-( what do you test here?
```diff
diff --git a/test/dyndns-test.xml b/test/dyndns-test.xml
index 16ad005..38af667 100644
--- a/test/dyndns-test.xml
+++ b/test/dyndns-test.xml
@@ -3,34 +3,5 @@
   <system>
   </system>
   <OPNsense>
-    <DynDNS>
-      <general>
-        <enabled>1</enabled>
-        <verbose>0</verbose>
-        <allowipv6nano>0</allowipv6nano>
-        <daemon_delay>300</daemon_delay>
-        <backend>opnsense</backend>
-      </general>
-      <accounts>
-        <account uuid="8e4627c4-21ff-4252-a331-3d1adee0a023">
-          <enabled>1</enabled>
-          <service>noip</service>
-          <protocol/>
-          <server/>
-          <username/>
-          <password>user</password>
-          <resourceId>pass</resourceId>
-          <hostnames>all.ddnskey.com</hostnames>
-          <wildcard>0</wildcard>
-          <zone/>
-          <checkip>web_noip-ipv4</checkip>
-          <checkip_timeout>10</checkip_timeout>
-          <force_ssl>1</force_ssl>
-          <ttl>300</ttl>
-          <interface>wan</interface>
-          <description>dyndns-description</description>
-        </account>
-      </accounts>
-    </DynDNS>
   </OPNsense>
 </opnsense>
```

**Review by zerwes — CHANGES_REQUESTED** [review]
test required, not a fake

**Review by zerwes — APPROVED** [review]
finally :+1:

**Comment by zerwes** [comment]
due to the 2 steps back and one forward commits I will perform a squash merge to main

**Pull request merged**